### DISCUSSION TITLE
Validation error when use alias in pydantic return model

### DISCUSSION THREAD

**Discussion by zihaooo** [fastapi/fastapi#7507] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Bing "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
import fastapi
import starlette.testclient
from pydantic import BaseModel, Field

app = fastapi.FastAPI()


class InModel(BaseModel):
    start_time: str = Field(alias='start_hms')


class OutModel(BaseModel):
    end_time: str = Field(alias='end_hms')


@app.post('/test', response_model=OutModel)
def test(query_body: InModel):
    print(query_body)  # start_time='14:00:00'
    print(query_body.dict())  # {'start_time': '14:00:00'}
    out_model = {'end_time': '12:00:00'}
    return out_model


with starlette.testclient.TestClient(app) as test_client:
    try:
        response = test_client.post('/test', json={'start_hms': '14:00:00'})
        print(response.json())
    except Exception as exc:
        print(exc)
```

### Description

A wonderful framework! Very happy to use it to coding. Here is a little problem that make me really confused. I can easily change the name for request body from public name to field name by add alias in pydantic model (just what i did to InModel in the code), but when I do the same thing to the return model (OutModel in the code), an error occurred.
* Run the code.
* It ends up with an exception: 
```
1 validation error for OutModel
response -> end_hms
  field required (type=value_error.missing).
```
* But I expected it to return `{"end_hms": "14:00:00"}`.

### Environment

* OS: Ubuntu 20.04
* FastAPI version: 0.58.0
* Python version: 3.8.2
* Pydantic version: 1.5.1
* Starlette version: 0.13.4

**Comment by MacMacky** [comment, marked as answer]
Hi @zihaooo you could do it like this.

```py
class OutModel(BaseModel):
    end_time: str = Field(alias='end_hms')

    class Config:
        # so we can use either the 'end_time' attribute or the alias attribute in this case 'end_hms'
        allow_population_by_field_name = True
```

**Comment by zihaooo** [comment]
@MacMacky Many thanks~

For someone has the same question: https://pydantic-docs.helpmanual.io/usage/model_config/

And I complete the example with @MacMacky 's code:

```Python
import fastapi
import starlette.testclient
from pydantic import BaseModel, Field

app = fastapi.FastAPI()


class InModel(BaseModel):
    start_time: str = Field(alias='start_hms')


class OutModel(BaseModel):
    end_time: str = Field(alias='end_hms')
    class Config:
        # so we can use either the 'end_time' attribute or the alias attribute in this case 'end_hms'
        allow_population_by_field_name = True

@app.post('/test', response_model=OutModel)
def test(query_body: InModel):
    print('Enter server side...')
    print(f'Query body: {query_body.dict()}')  # {'start_time': '14:00:00'}
    out_model = {'end_time': '20:00:00'}
    print(f'Return value: {out_model}')
    print('Making response...\n')
    return out_model


with starlette.testclient.TestClient(app) as test_client:
    try:
        print('Enter client side...')
        query_body = {'start_hms': '14:00:00'}
        print(f'Query body: {query_body}')
        print('Making request...\n')
        response = test_client.post('/test', json=query_body)
        print('Back to client side...')
        print(f'Return value: {response.json()}')
        print('Exit client side.\n')
    except Exception as exc:
        print(exc)
```
Output: 
```
Enter client side...
Query body: {'start_hms': '14:00:00'}
Making request...

Enter server side...
Query body: {'start_time': '14:00:00'}
Return value: {'end_time': '20:00:00'}
Making response...

Back to client side...
Return value: {'end_hms': '20:00:00'}
Exit client side.
```

**Comment by MacMacky** [comment]
@zihaooo You're welcome, man. Glad that I helped you.
Will you please close this issue. Thanks.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:

**Comment by nicktimko** [comment]
I also recently ran across this, which I kinda view as a problem. If we set up a class that has a `nice_internal_name` with a crappy external alias we're dealing with (maybe it's abstruse, maybe it's a keyword like `and`, `class`, etc.), internally, we still need to instantiate the class with the alias name, even though within the code we're using that alias to avoid dealing with it entirely.
```py
from pydantic import BaseModel, Field

class MyObj(BaseModel):
    class Config:
        populate_by_name = True  # if False (default), will cause exception on 'THIS LINE'

    nice_internal_name: str = Field(alias="fhqwhgads")

# pretend deserialize from external source
x = MyObj.model_validate_json('{"fhqwhgads": "homestar"}')

# repr of the class shows the nice_internal_name regardless...
x_repr = repr(x) 
assert x_repr == "MyObj(nice_internal_name='homestar')"

# but if we want to create a class instance internally, perhaps for testing, we 
# need to use the external name?! Only if populate_by_name is false-y.
# https://docs.python.org/3/library/functions.html#repr advises: 
#     For many types, this function makes an attempt to return a string that would 
#     yield an object with the same value when passed to eval();"
x_roundtrip = eval(x_repr)  # THIS LINE
assert x_roundtrip == x

# can serialize back to the external name
xj = x.model_dump_json(by_alias=True)
assert xj == '{"fhqwhgads":"homestar"}'
```

Strangely, the `repr`/`eval` cycle is broken when using the *default* `populate_by_name = False`, which feels kinda wrong. Needing to `eval` is slightly beside the point, which is: entirely avoiding the use of the alias/external name within the code. This feels wrong:

```py
x = MyObj(fhqwhgads="thing")
assert x.nice_internal_name == "thing"
```
